File: ts/src/storage/s3-backend.ts
import { AwsClient } from 'aws4fetch';
import type { RadStorageBackend } from './backend';

export interface S3Config {
  endpoint: string;
  bucket: string;
  accessKey: string;
  secretKey: string;
  region: string;
}

/**
 * S3-compatible storage backend.
 * Works with AWS S3, Cloudflare R2, Backblaze B2, iDrive, MinIO, etc.
 */
export class S3Backend implements RadStorageBackend {
  private client: AwsClient;
  private endpoint: string;
  private bucket: string;
  private bucketCreated: boolean = false;

  constructor(config: S3Config) {
    this.client = new AwsClient({
      accessKeyId: config.accessKey,
      secretAccessKey: config.secretKey,
      region: config.region,
      service: 's3',
    });
    this.endpoint = config.endpoint;
    this.bucket = config.bucket;
  }

  private async ensureBucket(): Promise<void> {
    if (this.bucketCreated) {
      return;
    }

    try {
      // Try to create the bucket (PUT with empty body)
      const url = `${this.endpoint}/${this.bucket}`;
      const res = await this.client.fetch(url, {
        method: 'PUT',
        body: '',
      });

      // Bucket creation succeeds with 200, 201, or 409 (already exists)
      if (res.ok || res.status === 409) {
        this.bucketCreated = true;
        return;
      }

      // If bucket creation failed, try a test PUT to see if bucket exists
      const testKey = `_test/${Date.now()}`;
      const testUrl = `${this.endpoint}/${this.bucket}/${testKey}`;
      const testRes = await this.client.fetch(testUrl, {
        method: 'PUT',
        body: 'test',
      });

      if (testRes.ok) {
        this.bucketCreated = true;
        // Clean up test file
        await this.client.fetch(testUrl, { method: 'DELETE' });
      }
    } catch (e) {
      // Bucket creation might have failed, but bucket might exist
      console.warn('Bucket initialization warning:', e);
    }
  }

  async put(key: string, data: string): Promise<void> {
    await this.ensureBucket();

    const url = `${this.endpoint}/${this.bucket}/${key}`;
    const res = await this.client.fetch(url, {
      method: 'PUT',
      body: data,
    });

    if (!res.ok) {
      const errorBody = await res.text();
      console.error(`S3 PUT failed for ${key}:`, res.status, res.statusText, errorBody.substring(0, 200));
      throw new Error(`S3 PUT failed: ${res.status} ${res.statusText}`);
    }
  }

  async get(key: string): Promise<string | null> {
    const url = `${this.endpoint}/${this.bucket}/${key}`;
    const res = await this.client.fetch(url);

    if (res.status === 404) {
      return null;
    }

    if (!res.ok) {
      // Handle bucket not found or other errors during initialization
      console.warn(`S3 GET warning for ${key}: ${res.status} ${res.statusText}`);
      return null;
    }

    return await res.text();
  }

  async list(prefix: string): Promise<string[]> {
    const url = `${this.endpoint}/${this.bucket}?list-type=2&prefix=${encodeURIComponent(prefix)}`;
    const res = await this.client.fetch(url);

    // Handle bucket not found or no objects - return empty array
    if (res.status === 404) {
      return [];
    }

    if (!res.ok) {
      // For other errors, log but return empty to allow initialization
      console.warn(`S3 LIST warning: ${res.status} ${res.statusText}`);
      return [];
    }

    const xml = await res.text();

    // Parse XML to extract keys
    // S3 ListObjectsV2 response format:
    // <ListBucketResult><Contents><Key>path/to/file</Key></Contents>...</ListBucketResult>
    const keys: string[] = [];
    const keyRegex = /<Key>([^<]+)<\/Key>/g;
    let match;

    while ((match = keyRegex.exec(xml)) !== null) {
      keys.push(match[1]);
    }

    return keys.sort();
  }

  async delete(key: string): Promise<void> {
    const url = `${this.endpoint}/${this.bucket}/${key}`;
    const res = await this.client.fetch(url, {
      method: 'DELETE',
    });

    if (!res.ok && res.status !== 404) {
      throw new Error(`S3 DELETE failed: ${res.status} ${res.statusText}`);
    }
  }
}

File: rust/src/storage/s3_backend.rs
use async_trait::async_trait;
use s3::Bucket;
use s3::creds::Credentials;
use s3::Region;
use super::backend::RadStorageBackend;

pub struct S3Config {
    pub endpoint: String,
    pub bucket: String,
    pub access_key: String,
    pub secret_key: String,
    pub region: String,
}

/// S3-compatible storage backend.
/// Works with AWS S3, Cloudflare R2, Backblaze B2, iDrive, MinIO, etc.
pub struct S3Backend {
    bucket: Box<Bucket>,
}

impl S3Backend {
    pub fn new(config: S3Config) -> Result<Self, String> {
        let credentials = Credentials::new(
            Some(&config.access_key),
            Some(&config.secret_key),
            None,
            None,
            None,
        ).map_err(|e| format!("Failed to create credentials: {}", e))?;

        let region = Region::Custom {
            region: config.region.clone(),
            endpoint: config.endpoint.clone(),
        };

        let bucket = Bucket::new(&config.bucket, region, credentials)
            .map_err(|e| format!("Failed to create bucket: {}", e))?;

        Ok(Self { bucket })
    }
}

#[async_trait]
impl RadStorageBackend for S3Backend {
    async fn put(&self, key: &str, data: &str) -> Result<(), String> {
        self.bucket
            .put_object(key, data.as_bytes())
            .await
            .map_err(|e| format!("S3 PUT failed: {}", e))?;

        Ok(())
    }

    async fn get(&self, key: &str) -> Result<Option<String>, String> {
        match self.bucket.get_object(key).await {
            Ok(response) => {
                let data = String::from_utf8(response.bytes().to_vec())
                    .map_err(|e| format!("Invalid UTF-8: {}", e))?;
                Ok(Some(data))
            }
            Err(e) => {
                let err_str = e.to_string();
                // Check if it's a 404 error
                if err_str.contains("404") || err_str.contains("NoSuchKey") {
                    Ok(None)
                } else {
                    Err(format!("S3 GET failed: {}", e))
                }
            }
        }
    }

    async fn list(&self, prefix: &str) -> Result<Vec<String>, String> {
        let results = self.bucket
            .list(prefix.to_string(), None)
            .await
            .map_err(|e| format!("S3 LIST failed: {}", e))?;

        let mut keys = Vec::new();
        for list in results {
            for obj in list.contents {
                keys.push(obj.key);
            }
        }

        keys.sort();
        Ok(keys)
    }

    async fn delete(&self, key: &str) -> Result<(), String> {
        self.bucket
            .delete_object(key)
            .await
            .map_err(|e| format!("S3 DELETE failed: {}", e))?;

        Ok(())
    }
}
